src/main/java/org/sqlite/core/NativeDB.c

```diff
@@ -429,6 +434,26 @@ void xFinal(sqlite3_context *context)
     (*env)->DeleteGlobalRef(env, *func);
 }
 
+int xCompare(void* context, int len1, const void* str1, int len2, const void* str2)
+{
+    static jmethodID mth = 0;
```

Comment:
What if we register two or more Collation functions to the connection? I guess only the first one wins and collation functions created later will never be used.

Reply:
I have added a test to validate that it works, please check the latest commit.

The custom collation works a bit differently than the UDF. [Here](https://www.sqlite.org/c3ref/create_collation.html) you can read:

> The fourth argument, pArg, is an application data pointer that is passed through as the first argument to the collating function callback.

[This](https://github.com/gotson/sqlite-jdbc/blob/884288d946bfc1204b4db1567df91ca81436464d/src/main/java/org/sqlite/core/NativeDB.c#L1360) is what we pass, it's a struct containing the pointer to the custom function

> The collating function callback is invoked with a copy of the pArg application data pointer

In the callback, [we can access](https://github.com/gotson/sqlite-jdbc/blob/884288d946bfc1204b4db1567df91ca81436464d/src/main/java/org/sqlite/core/NativeDB.c#L441) the previously defined `coll` struct. [Here](https://github.com/gotson/sqlite-jdbc/blob/884288d946bfc1204b4db1567df91ca81436464d/src/main/java/org/sqlite/core/NativeDB.c#L454) we call the method on that particular object, so no conflict can arise.

---

I have some doubts on whether the memory is actually freed for those `coll` structs. When I wrote this code, i tried to find how the UDF data pointers are freed, but couldn't find anything. Today i found that it's done [here](https://github.com/gotson/sqlite-jdbc/blob/884288d946bfc1204b4db1567df91ca81436464d/src/main/java/org/sqlite/core/NativeDB.c#L1386). So there's probably something to do to properly free the memory for the custom collation functions too.